protocol.md

```diff
@@ -726,6 +727,140 @@ Upload-Length: 11
 Upload-Concat: final;/files/a /files/b
 ```
 
+### Challenge
+
+With this extension, Clients can authenticate follow-up requests to a [Creation](#creation) or [Creation With Upload](#creation-with-upload) 
+`POST` request using a high-entropy cryptographic random shared secret - and challenges
+based thereon.
+
+If the Server supports this extension, it MUST add `challenge` to the `Tus-Extension` header.
+
+#### Headers
+
+##### Upload-Secret
+
+The `Upload-Secret` MUST be a high-entropy cryptographic random string, consisting
+of 48 to 256 characters from the [base64 alphabet](https://tools.ietf.org/html/rfc4648#section-4).
+That's 36 to 192 bytes encoded as base64.
+
+The `Upload-Secret` header MUST NOT be included with anything but `POST` requests
+that create a new upload resource. Servers receiving an `Upload-Secret` header with
+any other request MUST respond with a `400 Bad Request` status.
+
+##### Upload-Challenge
+
+If an `Upload-Secret` header was sent with the `POST` request that created
+the upload resource, Clients MUST include an `Upload-Challenge` header with every
+subsequent request targeting that upload resource.
+
+The value of `Upload-Challenge` is defined as follows:
+
+```
+Upload-Challenge = SHA256([Upload-Secret] + SHA256(`[HTTP Method]` + [Upload-Offset] + [Upload-Secret] + [Content-Length]))
```

Comment:
Is there a benefit from having nested hash functions here or could it we just use following:

> Upload-Challenge = SHA256([Upload-Secret] + [HTTP Method] + [Upload-Offset])

Reply:
It requires _slightly_ more computation, so would slow down brute force attacks. But it doesn't offer other benefits as far as I can tell. I think the simplified version should be just as secure.

Towards brute force attacks, the addition of `Upload-Challenge-Salt` (maybe just as `Upload-Salt`) suggested by @smatsson should be more effective.

Reply:
It it does not bring much extra security I think we should make it easier to compute. Crypto is hard enough for people anyway :)

Reply:
Agreed!